helix-view/src/editor.rs

```diff
@@ -398,18 +398,105 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
+
+    pub fn close_document(&mut self, id: DocumentId, force: bool) -> anyhow::Result<()> {
+        if !self.documents.contains_key(&id) {
+            anyhow::bail!("document does not exist");
+        }
+
+        log::error!("remove: {:?}", &id);
+        let doc = &self.documents[&id];
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let old_focus = self.tree.focus;
+        let view_ids = self
+            .tree
+            .traverse()
+            .filter(|(_, view)| view.doc == id)
+            .map(|(id, _)| id)
+            .collect::<Vec<_>>();
+
+        // NOTE: issues are:
+        // * have to set the tree focus in order for switch to work properly (it
+        // wants to switch to the document in the currently focused document,
+        // which doesn't work well for modifying *other* views)
+        //   -> maybe make switch take a viewid explicitly, get other stuff from that?
```

Comment:
You probably don't want to use `editor.switch` since it'll try to push a jump to a doc you're removing. I would extract https://github.com/helix-editor/helix/blob/f78c07db791fa25dd20bb5efd22cc8065d964874/helix-view/src/editor.rs#L271-L286 into a new method and use that (add a `current: ViewId` param so we avoid changing focus)